From: Cesar Strauss <cstrauss@not2005.cea.inpe.br>
Date: Sat, 9 Oct 2010 21:37:26 +0000 (-0300)
Subject: Check for the _spawnvp function and use it if present.
X-Git-Url: http://git.gpleda.org/?p=pcb.git;a=commitdiff_plain;h=02fa7cf6646ad4f16383c91e53784986f4d8d15d;hp=70786c0c18c8d2c4626d429068e88a4601a21d6d

Check for the _spawnvp function and use it if present.
---

diff --git a/configure.ac b/configure.ac
index a220a82..abdd8a1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -662,6 +662,9 @@ AC_CHECK_FUNCS(popen)
 AC_CHECK_FUNCS(realpath canonicalize_file_name)
 libiberty_NEED_DECLARATION(canonicalize_file_name)
 
+# for pcb_spawnvp in action.c on Windows
+AC_CHECK_FUNCS(_spawnvp)
+
 AC_HEADER_STDC
 AC_CHECK_HEADERS(limits.h locale.h string.h sys/types.h regex.h pwd.h)
 AC_CHECK_HEADERS(sys/socket.h netinet/in.h netdb.h sys/param.h sys/times.h sys/wait.h)
diff --git a/src/action.c b/src/action.c
index 0e035a6..88907af 100644
--- a/src/action.c
+++ b/src/action.c
@@ -7287,6 +7287,13 @@ ActionExecCommand (int argc, char **argv, int x, int y)
 static int
 pcb_spawnvp (char **argv)
 {
+#ifdef HAVE__SPAWNVP
+  int result = _spawnvp (_P_WAIT, argv[0], argv);
+  if (result == -1)
+    return 1;
+  else
+    return 0;
+#else
   int pid;
   pid = fork ();
   if (pid < 0)
@@ -7308,6 +7315,7 @@ pcb_spawnvp (char **argv)
       wait (&rv);
     }
   return 0;
+#endif
 }
 
 /* ---------------------------------------------------------------- */
